File: src/sound/Sound.cpp
/*
 * Sound.cpp
 *
 *  Created on: 2018年1月1日
 *      Author: neo
 */

#include "Sound.h"
#include "script/NeoScript.h"

namespace NeoGame
{
SINGLETON_SOURCE(Sound)

Sound::Sound() :
		m_audio_manager(0), m_audio_id_counter(0)
{

}

bool Sound::Init()
{
	std::string deviceName =
			NeoScript::getInstance()->GetScriptGlobalValue_string(
					"ApplicationSettings.SoundDevice");
	if (m_audio_manager)
	{
		m_audio_manager->shutDown();
		cAudio::destroyAudioManager(m_audio_manager);
	}
	m_device_names.clear();
	// get device list
	cAudio::IAudioDeviceList* device_list = cAudio::createAudioDeviceList();
	unsigned int count = device_list->getDeviceCount();
	m_default_device = std::string(device_list->getDefaultDeviceName().c_str());
	bool in_list = false;
	for (unsigned int i = 0; i < count; i++)
	{
		std::string d = device_list->getDeviceName(i);
		m_device_names.push_back(d);
		if (d == deviceName)
		{
			in_list = true;
		}
	}

	//init manager
	bool success = false;
	if (deviceName.empty() || !in_list)
	{
		m_audio_manager = cAudio::createAudioManager();
		success = true;
	}
	else
	{
		m_audio_manager = cAudio::createAudioManager(false);
		success = m_audio_manager->initialize(deviceName.c_str());
	}

	CAUDIO_DELETE device_list;
	if (success)
	{
		m_listener.setListener(m_audio_manager->getListener());
	}
	return success;
}

bool Sound::setAudioDevice(std::string deviceName)
{
	if (!deviceName.empty() && m_audio_manager)
	{
		m_audio_manager->shutDown();
		return m_audio_manager->initialize(deviceName.c_str());
	}
	return false;
}

std::vector<std::string>& Sound::getDeviceList()
{
	return m_device_names;
}

void Sound::setMasterVolume(float volume)
{
	if (m_audio_manager)
	{
		m_audio_manager->setMasterVolume(volume);
	}
}

float Sound::getMasterVolume()
{
	if (m_audio_manager)
	{
		return m_audio_manager->getMasterVolume();
	}
	return 0;
}

cAudio::IAudioManager* Sound::getAudioManager() const
{
	return m_audio_manager;
}

const std::string& Sound::getDefaultDevice()
{
	return m_default_device;
}

AudioListener& Sound::getListener()
{
	return m_listener;
}

AudioSource* Sound::Play2DSound(std::string path, bool loop)
{
	if (m_audio_manager)
	{
		AudioSource* audio = CreateInstantAudioSource("", path.data(), true);
		if (audio)
		{
			audio->play2d(loop);
			return audio;
		}
	}
	return nullptr;
}

AudioSource* Sound::Play3DSound(std::string path,
		const irr::core::vector3df& pos, bool loop)
{
	if (m_audio_manager)
	{
		AudioSource* audio = CreateInstantAudioSource("", path.data(), true);
		if (audio)
		{
			audio->play3d(pos, 1.0, loop);
			return audio;
		}
	}
	return nullptr;
}

AudioSource* Sound::CreateInstantAudioSource(const char* name,
		const char* filename, bool stream/* = false*/)
{
	if (m_audio_manager)
	{
		if (m_instant_audio_soureces.find(filename)
				== m_instant_audio_soureces.end())
		{
			m_instant_audio_soureces[filename] = AudioPool();
		}
		AudioPool& pool = m_instant_audio_soureces[filename];
		cAudio::IAudioSource* audio = m_audio_manager->create(name, filename,
				stream);
		if (audio)
		{
			std::shared_ptr<AudioSource> audio_source = pool.Create();
			if (!audio_source->isInitialized())
			{
				audio_source->Init(audio, filename);
				InstantSoundEventHandler* handler =
						new InstantSoundEventHandler(audio_source);
				m_instant_audio_event_handlers[filename] = std::shared_ptr<
						InstantSoundEventHandler>(handler);
				audio->registerEventHandler(handler);
			}
			return audio_source.get();
		}
	}
	return nullptr;
}

AudioSource* Sound::CreateAudioSource(const char* name, const char* filename,
		bool stream)
{
	cAudio::IAudioSource* audio = m_audio_manager->create(name, filename,
			stream);
	AudioSource* audio_source = new AudioSource(audio, filename);
	if (not audio_source)
	{
		return nullptr;
	}

	m_audio_sources[m_audio_id_counter] = audio;
	audio_source->setId(m_audio_id_counter);
	m_audio_id_counter++;
	return audio_source;
}

void Sound::ReleaseAudioSource(AudioSource* audio)
{
	auto iter = m_audio_sources.find(audio->getId());
	if (iter != m_audio_sources.end())
	{
		m_audio_manager->release(audio->getAudioSource());
		m_audio_sources.erase(iter);
	}
}

void Sound::Clear()
{
	m_instant_audio_soureces.clear();
	m_audio_sources.clear();
	m_audio_manager->releaseAllSources();
	m_instant_audio_event_handlers.clear();
	m_audio_id_counter = 0;
}

void Sound::RecycleAudioSource(std::shared_ptr<AudioSource> audio)
{
	if (m_instant_audio_soureces.find(audio->getFilePath())
			!= m_instant_audio_soureces.end())
	{
		m_instant_audio_soureces[audio->getFilePath()].Recycle(audio);
	}
}

void Sound::StopAllSounds()
{
	if (m_audio_manager)
	{
		m_audio_manager->stopAllSounds();
	}
}

Sound::~Sound()
{
	if (m_audio_manager)
	{
		m_audio_manager->stopAllSounds();
		Clear();
		cAudio::destroyAudioManager(m_audio_manager);
	}
}

} /* namespace NeoGame */

File: script/game/sound.lua
require(DIR_DATA.."game/sound_map")
Sound={
	sound_engine=NeoGame.Sound:getInstance(),
	bgm={},
	effects={},
}

function Sound:StopAllSounds()
	self.sound_engine:StopAllSounds()
end

function Sound:Clear()
	self.sound_engine:Clear()
end

function Sound:Play2D(sound_key, loop, sound_type)
	if sound_key and sound_map[sound_key] then
		sound_type =sound_type or "bgm"
		loop = loop or false
		local volume = 1
		if sound_type == "bgm" then
			volume = ApplicationSettings.SoundVolume.bgm/100
		elseif sound_type == "effect" then
			volume = ApplicationSettings.SoundVolume.effect/100
		end
		local s = self.sound_engine:Play2DSound(DIR_SOUND..sound_map[sound_key],loop)
		s:setVolume(volume)
		return s
	end
	return nil
end

function Sound:Play3D(sound_key, position,loop, sound_type)
	if sound_key and sound_map[sound_key] then
		sound_type =sound_type or "effect"
		loop=loop or false
		local volume = 1
		if sound_type == "bgm" then
			volume = ApplicationSettings.SoundVolume.bgm/100
		elseif sound_type == "effect" then
			volume = ApplicationSettings.SoundVolume.effect/100
		end
		position = position or irr.core.vector3df:new_local()
		local s = self.sound_engine:Play3DSound(DIR_SOUND..sound_map[sound_key],position,loop)
		s:setVolume(volume)
		return s
	end
	return nil
end

function Sound:CreateAudioSource( sound_type,sound_key, sound_name, stream )
	if sound_key and sound_map[sound_key] then
		sound_type =sound_type or "effect"
		sound_name = sound_name or ""
		if stream == nil then
			stream=true
		end
		local volume = 1
		if sound_type == "bgm" then
			volume = ApplicationSettings.SoundVolume.bgm/100
		elseif sound_type == "effect" then
			volume = ApplicationSettings.SoundVolume.effect/100
		end
		local s = self.sound_engine:CreateAudioSource(sound_name,DIR_SOUND..sound_map[sound_key],stream)
		s:setVolume(volume)
		return s
	end
	return nil
end

function Sound:ReleaseAudioSource( audio_source )
	self.sound_engine:ReleaseAudioSource(audio_source)
end

function Sound:setMasterVolume( volume )
	volume = volume or 1
	self.sound_engine:setMasterVolume(volume)
end

function Sound:getListener()
	return self.sound_engine:getListener()
end

function Sound:getDeviceList( )
	return self.sound_engine:getDeviceList()
end